### ISSUE TITLE
[Persistence] refactor ReadContext and RWContext interfaces

### ISSUE THREAD

**Issue by deblasis** [pokt-network/pocket#406]
## Objective

Improve `ReadContext` and `RWContext` interfaces so that their usage is simpler and more semantic.

## Origin Document

While touching parts of the code I haven't previously interacted with I have noticed some oddities that were clarified in code review.

For context: https://github.com/pokt-network/pocket/pull/378#discussion_r1044914839

## Goals

- Make sure that the ReadContext constructor/initializer doesn't take an height argument
- Make sure that the ReadContext related methods have a height argument
- Make sure that the RWContext constructor/initializer .... **TBD**
- Make sure that the RWContext related methods don't take the height (since it's derived from the instance of the context itself)

## Deliverable

- [ ] Updated ReadContext interface
- [ ] Updated ReadContext mocks & mock configs
- [ ] Updated RWContext interface
- [ ] Updated RWContext mocks & mock configs
- [ ] Updated tests

## Non-goals / Non-deliverables

- Changes in logic 

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: [@deblasis ]
**Co-Owners**: [@Olshansk]

**Comment by deblasis** [comment]
@Olshansk 

One last thing about this, to quote the relevant part of our [conversation](https://github.com/pokt-network/pocket/pull/378#discussion_r1050275785):

![image](https://user-images.githubusercontent.com/29378614/208121787-760a4bc2-8c36-4e80-9037-91c5ba17ddeb.png)

> "I am now going to be making modifications to height h - acquire read/write lock on the db"

I get it, my question would then be:

Isn't `h` always = `currentHeight`? I thought that once a block is committed we can't go back and change things, ergo:

`h` is never < `currentHeight`

I can only think about rollbacks as an exception to this but I would implement them with a completely separate interface to keep things cleanly separated.

If we don't know at this point what is the best way to address this, I'll figure it out.

**Comment by Olshansk** [comment]
That is SUPER interesting @deblasis! I see what you're saying and I agree. 

--- 

Applying blocks during state sync (staked actors and full nodes) - We can download data in parallel but processing (i.e. block application) has to be sequential. **tl;dr Makes sense that h = latest**.

Applying blocks during consensus (validators only) - you're only writing to the current height. **tl;dr Makes sense that h = latest**.

Rollbacks - I think you'd still be able to (ideally should) leverage the same interface. Get the current block (or read from the SQL DB), and undo the state transition at that height so you go back, safely, to a block-1 .

I don't see any gaps in your intuition at the moment and the approach makes sense to me, seems simpler and exactly what we need. LFG

**Comment by deblasis** [comment]
🫡

**Comment by adshmh** [comment]
I came across this issue while working on #803.
Removing height from `NewReadContext` input and passing it to specific methods seems a better choice. This could prevent confusion when retrieving the value of a parameter for a height less than the current height, e.g. when getting the App Token sessions multiplier for the starting height of a session.

**Comment by Olshansk** [comment]
> I came across this issue while working on #803. Removing height from `NewReadContext` input and passing it to specific methods seems a better choice. This could prevent confusion when retrieving the value of a parameter for a height less than the current height, e.g. when getting the App Token sessions multiplier for the starting height of a session.

That's the plan!

**Comment by Olshansk** [comment]
Just pointing out that @bryanchriswhite raised this point again in #812.

![Screenshot 2023-06-20 at 1 31 04 PM](https://github.com/pokt-network/pocket/assets/1892194/8af99961-2434-454d-9d02-d6c2cba7832f)

@dylanlott Could you look into prioritizing this in your next PR?

Also CCing @h5law just so he's aware of this as well.
